### PR TITLE
Add AI PR Reviewer workflow configuration

### PR DESCRIPTION
**Pull request by adillpickles**
## Description
Integrated Codium PR-Agent as a replacement for GitHub Copilot's PR review functionality (when rate limited). This setup uses a dedicated GitHub Action and the DeepSeek V4 Pro model to provide automated code reviews, PR summaries, and inline code suggestions.

## Related User Story / Issue
Not related to a user story or issue - Infrastructure Improvement

## Type of Change
- [X] New feature
- [ ] Bug fix
- [ ] Backend / API change
- [ ] UI change
- [ ] Database change
- [X] Documentation/Configuration
- [ ] Refactor

## How Has This Been Tested?
- [ ] GitHub Actions / CI tests passed
- [ ] Manual testing performed
- [X] Not tested locally

### Test Notes
This is the initial setup. Testing will occur post-merge or during the PR process once the Action is triggered for the first time. I have added the DEEPSEEK_API_KEY to the repository secrets manually.

## UI Changes (if applicable)
- [X] No UI changes
- [ ] UI updated (add screenshots if needed)

## Notes for Reviewers
Once merged, the bot should automatically comment on new PRs. If it doesn't, we'll need to debug the YAML configuration in the next pass.

## Checklist
- [ ] Linked to a user story or issue
- [ ] Code builds / tests pass in CI
- [X] Ready for review

### CODE REVIEW

File: .github/workflows/pr_agent.yml [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    types: [created, edited]
+
+jobs:
+  pr_agent:
+    if: ${{ github.event.sender.type != 'Bot' && ((github.event_name == 'pull_request' && github.event.pull_request.draft == false) || (github.event_name == 'issue_comment' && github.event.issue.pull_request)) }}
+    runs-on: ubuntu-latest
+    permissions:
+      issues: write
+      pull-requests: write
+      contents: read
```
Comment (zakariac22) [inline .github/workflows/pr_agent.yml:16]: This was a good safeguarding measure to ensure it is unable to write to repository contents

File: .github/workflows/pr_agent.yml [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    steps:
+      - name: PR Agent
+        uses: Codium-ai/pr-agent@main
```
Comment (zakariac22) [inline .github/workflows/pr_agent.yml]: I would tie it to a specific commit SHA that it has supply chain protection

File: .github/workflows/pr_agent.yml [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+          pr_description.include_generated_by_header: "false"
+          pr_description.collapsible_file_list: "true"
+          pr_description.collapsible_file_list_threshold: "1"
+          pr_description.inline_file_summary: "table"
+          pr_description.extra_instructions: "Write a concise Copilot-style pull request overview. Start with the heading \"## Pull request overview\", then one short paragraph, then a \"**Changes:**\" list with at most three bullets. Keep the per-file walkthrough brief and avoid repeating the original PR body or template text."
+          pr_reviewer.num_max_findings: "12"
+          pr_reviewer.require_security_review: "true"
+          pr_reviewer.require_tests_review: "true"
+          pr_reviewer.extra_instructions: "Review like a staff engineer: product context, correctness, edge cases, security, tests, and maintainability. Prefer concise, actionable inline feedback where GitHub allows it; use the summary for themes and risk."
+          pr_code_suggestions.commitable_code_suggestions: "true"
```
Comment (zakariac22) [inline .github/workflows/pr_agent.yml:59]: I would set this to false to ensure this can only give suggestions, never make commits in the repo

File: .github/workflows/pr_agent.yml [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+on:
+  pull_request:
+    types: [opened, reopened, ready_for_review, synchronize]
+  issue_comment:
+    types: [created, edited]
+
+jobs:
+  pr_agent:
+    if: ${{ github.event.sender.type != 'Bot' && ((github.event_name == 'pull_request' && github.event.pull_request.draft == false) || (github.event_name == 'issue_comment' && github.event.issue.pull_request)) }}
+    runs-on: ubuntu-latest
```
Comment (zakariac22) [inline .github/workflows/pr_agent.yml:11]: I would add another line under pr_agent: which is timeout-minutes: 10, for example, to prevent excessively long runs. This can be adjusted for what is reasonable, but if not configured, it will take a default timeout.

File: .pr_agent.toml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+require_tests_review = true
+extra_instructions = """
+Review like a staff engineer: product context, correctness, edge cases, security, tests, and
+maintainability. Prefer concise, actionable inline feedback where GitHub allows it; use the
+summary for themes and risk.
+"""
+
+[pr_code_suggestions]
+# Committable, Copilot-style suggestion blocks in the diff when the provider supports it.
+commitable_code_suggestions = true
```
Comment (zakariac22) [inline .pr_agent.toml:47]: I would set this to false even though the workflow is already recommended to enforce this.

File: .github/workflows/pr_agent.yml (hunk 2 above)
Comment (adillpickles) [reply .github/workflows/pr_agent.yml]: Updated

File: .github/workflows/pr_agent.yml (hunk 3 above)
Comment (adillpickles) [reply .github/workflows/pr_agent.yml:59]: Keeping this enabled because PR Agent only creates manual GitHub suggestion blocks. it cannot commit by itself. That matches the Copilot style workflow goal.

File: .github/workflows/pr_agent.yml (hunk 4 above)
Comment (adillpickles) [reply .github/workflows/pr_agent.yml:11]: Added a 30 minute timeout instead of 10 since the free endpoint can queue or respond slowly on larger PRs.

File: .pr_agent.toml (hunk 5 above)
Comment (adillpickles) [reply .pr_agent.toml:47]: Keeping this enabled because PR Agent only creates manual GitHub suggested change blocks. it does not auto commit. This matches the Copilot-style review workflow I was going for, where reviewers can optionally apply a suggestion. However I highly recommend against blindly pressing commit suggestion and still writing the code and using the suggestion as inspiration.

File: .github/workflows/pr_agent.yml [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+              return;
+            }
+
+            const body = describeComment.body;
+            const titleBlock = (body.match(/^## Title\s+([\s\S]*?)\n___/m)?.[1] || '').trim();
+            const title = titleBlock
+              .split(/\r?\n/)
+              .map((line) => line.trim())
+              .filter((line) => line && !/^####\s+\(Describe updated until commit/i.test(line))
+              .pop() || context.payload.pull_request.title;
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply .github/workflows/pr_agent.yml]: Valid suggestion though minor. Glad this is working though.

File: .github/workflows/pr_agent.yml [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            });
+
+            const fileRows = files
+              .slice(0, 20)
+              .map((file) => {
+                const description = file.status === 'removed'
+                  ? 'Removed file.'
+                  : `${file.status.charAt(0).toUpperCase()}${file.status.slice(1)} file with ${file.changes} changed lines.`;
+                return `| \`${file.filename.replaceAll('|', '\\|')}\` | ${description} |`;
+              });
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply .github/workflows/pr_agent.yml:136]: Good suggestion.

File: .github/workflows/pr_agent.yml [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            Reviewed the changed files and generated inline comments where applicable.
+
+            <details>
+            <summary>Show a summary per file</summary>
+
+            | File | Description |
+            | ---- | ----------- |
+            ${fileRows.join('\n')}${fileSummary}
+
+            </details>`;
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply .github/workflows/pr_agent.yml]: Very valid.

File: .github/workflows/pr_agent.yml [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+              .map((line) => line.replace(/^[-*]\s+/, '').trim())
+              .filter(Boolean)
+              .slice(0, 3);
+
+            const files = await github.paginate(github.rest.pulls.listFiles, {
+              owner,
+              repo,
+              pull_number: issue_number,
+              per_page: 100,
+            });
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply .github/workflows/pr_agent.yml:127]: Valid edge case, but this repo/class project will not have PRs anywhere near 3000 changed files so I’m leaving this as is to avoid over engineering the normalizer.

File: .github/workflows/pr_agent.yml [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+            const describeComment = comments
+              .filter((comment) =>
+                comment.user?.type === 'Bot' &&
+                comment.body?.startsWith('## Title') &&
+                comment.body?.includes('### **Description**')
+              )
+              .sort((a, b) => new Date(b.updated_at) - new Date(a.updated_at))[0];
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply .github/workflows/pr_agent.yml:99]: For this class project scope, I’m going to keep this as is rather than adding more workflow script complexity.

File: .github/workflows/pr_agent.yml (hunk 3 above)
Comment (zakariac22) [reply .github/workflows/pr_agent.yml:59]: This is a good point. For future reference, I would still make the suggested edits by making a commit yourself

**Review by zakariac22 — APPROVED** [review]
The necessary changes were made, and others left as-is for good reason.
